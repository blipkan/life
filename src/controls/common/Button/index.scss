.button {
  background-color: #5a3472;
  padding: 0 16px;
  border-radius: 4px;
  height: 36px;
  cursor: pointer;
  justify-content: center;
  align-items: center;
  display: flex;
  user-select: none;
  font-size: 14px;
  font-weight: 500;
}
