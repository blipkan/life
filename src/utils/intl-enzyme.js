import React from 'react'
import { IntlProvider } from 'react-intl'
import flatten from 'flat'

const defaultLocale = 'ru'
const locale = defaultLocale

const messages = require(`../../src/messages/${locale}.json`)

export const withIntl = component => {
  return (
    <IntlProvider locale={locale} messages={flatten(messages)}>
      {React.cloneElement(component)}
    </IntlProvider>
  )
}
