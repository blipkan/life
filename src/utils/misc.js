import keys from 'lodash/keys'
import isObject from 'lodash/isObject'

export const makeId = (length = 16) =>
  parseInt(
    Math.ceil(Math.random() * Date.now())
      .toPrecision(length)
      .toString()
      .replace('.', '')
  )

export const uniqArray = (array = []) => [...new Set(array)]

export const compactArray = (array = []) => array.filter(e => e)

export const consoleMessage = (message, messageType) => {
  const consoleMethod = ['error', 'warn', 'log', 'info'].includes(messageType)
    ? messageType
    : 'error'
  console[consoleMethod](message)
}

export const checkKey = (obj, key, errorMessageType, errorMessage) => {
  if (!isObject(obj)) {
    const message =
      errorMessage || `${JSON.stringify(obj)} cannot have key '${key}'`
    consoleMessage(message, errorMessageType)
    return
  }

  let result
  try {
    result = obj[key]
    if (!result) {
      const message =
        errorMessage ||
        `invalid key '${key}' for object: ${JSON.stringify(
          obj
        )}. Available keys: ${keys(obj)}`
      consoleMessage(message, errorMessageType)
    }
  } catch (e) {
    consoleMessage(e.message, errorMessageType)
  }

  return result
}

export const getIntlText = (formatMessage, messages = {}, key) => {
  const message = messages[key]
  return message ? formatMessage(message) : key
}
