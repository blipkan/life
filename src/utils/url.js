/**
 * @description get array from url search parameter
 * @param {string} paramName  parameter name
 * @returns {Array<string>}
 */
export const getValueFromUrlSearch = paramName => {
  const url = new URL(location.href)
  return url.searchParams.get(paramName)
}
