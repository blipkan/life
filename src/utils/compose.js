import { compactArray, uniqArray } from './misc'

export const Delimiter = {
  CLASSNAME: '__',
  MODIFIER: '_',
  ID: '_'
}

const createArrayCompositor = (
  base,
  delimiter = Delimiter.CLASSNAME,
  whiteSpaceAsDelimiter
) => {
  const strings = Array.isArray(base) ? uniqArray(compactArray(base)) : [base]

  const roots = !whiteSpaceAsDelimiter
    ? strings
    : strings.reduce((acc, string) => {
        return acc.concat(...string.split(' '))
      }, [])

  return (...siffixes) => {
    const path = siffixes.join(delimiter)
    return roots.map(
      id => `${id || ''}${id && path ? delimiter : ''}${path || ''}`
    )
  }
}

const createCompositor = (base, delimiter = Delimiter.CLASSNAME) => {
  const arrayCompositor = createArrayCompositor(base, delimiter)
  return (...siffixes) => arrayCompositor(...siffixes).join(' ')
}

const createCompositors = (
  className,
  id,
  {
    classNameDelimiter = Delimiter.CLASSNAME,
    idDelimiter = Delimiter.ID,
    modifierDelimiter = Delimiter.MODIFIER
  } = {}
) => {
  const classNameArrayCompositor = createArrayCompositor(
    className,
    classNameDelimiter,
    true
  )
  const classNameCompositor = (...siffixes) =>
    classNameArrayCompositor(...siffixes).join(' ')

  const modifierCompositor = (modifier, ...siffixes) => {
    const paths = classNameArrayCompositor(...siffixes)
    return paths
      .map(
        path =>
          `${path || ''}${
            modifier && path ? modifierDelimiter : ''
          }${modifier || ''}`
      )
      .join(' ')
  }

  const classNameAndModifierCompositor = (modifier, ...siffixes) => {
    const classNameString = classNameCompositor(...siffixes)
    const modifierString = modifierCompositor(modifier, ...siffixes)
    return `${classNameString} ${modifierString}`
  }

  const idCompositor = createCompositor(id, idDelimiter)

  return {
    id: idCompositor,
    className: classNameCompositor,
    modifier: modifierCompositor,
    classNameAndModifier: classNameAndModifierCompositor
  }
}

export const createCompose = (props = {}, defaults = {}) => {
  const { className, domId } = props
  return createCompositors(
    [defaults.className, className],
    domId || defaults.domId
  )
}
