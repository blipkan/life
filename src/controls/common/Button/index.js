import React from 'react'
import * as PropTypes from 'prop-types'

import { createCompose } from 'utils/compose'

import './index.scss'

export const Defaults = {
  className: 'button',
  domId: 'button'
}

const Button = props => {
  const { caption, hint, onClick } = props

  const compose = createCompose(props, Defaults)

  const doClick = e => {
    e.stopPropagation()
    if (typeof onClick === 'function') onClick(e)
  }

  return (
    <div
      className={compose.className()}
      id={compose.id()}
      title={hint || caption}
      onClick={doClick}
    >
      {caption}
    </div>
  )
}

Button.propTypes = {
  domId: PropTypes.string,
  className: PropTypes.string,
  caption: PropTypes.string,
  hint: PropTypes.string,
  onClick: PropTypes.func
}

export default Button
