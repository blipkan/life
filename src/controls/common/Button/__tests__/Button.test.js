/* eslint-env jest */

import React from 'react'
import { mount, shallow, render } from 'enzyme'

import Button, { Defaults } from '../index'
import { createCompose } from '../../../../utils/compose'

describe('Button', () => {
  const mockEventClick = { stopPropagation: jest.fn() }

  const DOMID = 'DOMID'
  const CLASSNAME = 'CLASSNAME'
  const CAPTION = 'CAPTION'
  const HINT = 'HINT'
  const onClick = jest.fn(e => e)

  const Properties = {
    className: CLASSNAME,
    domId: DOMID,
    caption: CAPTION,
    hint: HINT,
    onClick
  }

  it('render component', () => {
    const component = shallow(<Button />)
    expect(component).toMatchSnapshot()
  })

  it('render props', () => {
    const props = { ...Properties }
    const compose = createCompose(props, Defaults)

    const component = shallow(<Button {...props} />)

    expect(component).toMatchSnapshot()
    expect(component.prop('title')).toBe(props.hint)
    expect(component.prop('children')).toBe(props.caption)
    expect(component.prop('id')).toBe(compose.id())
    expect(component.prop('className')).toBe(compose.className())
  })

  it('handle empty hint (show caption)', () => {
    const props = { ...Properties, hint: null }
    const component = shallow(<Button {...props} />)
    const expectedHint = props.caption

    expect(component.prop('title')).toBe(expectedHint)
  })

  it('render empty caption', () => {
    const props = { ...Properties, caption: undefined }
    const component = shallow(<Button {...props} />)

    expect(component.prop('children')).not.toBe(expect.anything())
  })

  it('render composite className', () => {
    const props = { ...Properties, className: 'AAA BBB' }
    const component = shallow(<Button {...props} />)
    const compose = createCompose(props, Defaults)

    expect(component.prop('className')).toBe(compose.className())
  })

  it('call onClick', () => {
    const props = { ...Properties }
    const component = shallow(<Button {...props} />)
    component.simulate('click', mockEventClick)

    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
