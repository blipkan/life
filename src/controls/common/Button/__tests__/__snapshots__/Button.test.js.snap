// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render component 1`] = `
<div
  className="button"
  id="button"
  onClick={[Function]}
/>
`;

exports[`Button render props 1`] = `
<div
  className="button CLASSNAME"
  id="DOMID"
  onClick={[Function]}
  title="HINT"
>
  CAPTION
</div>
`;
